grid-client/deployer/deployment_deployer.go

```diff
@@ -29,102 +30,141 @@ func NewDeploymentDeployer(tfPluginClient *TFPluginClient) DeploymentDeployer {
+	}
+
+	dl.NodeDeploymentID, err = d.deployer.Deploy(
+		ctx, dl.NodeDeploymentID,
+		map[uint32]gridtypes.Deployment{dl.NodeID: dlsPerNodes[dl.NodeID][0]},
+		map[uint32]*uint64{dl.NodeID: dl.SolutionProvider},
+	)
 
 	// update deployment and plugin state
 	// error is not returned immediately before updating state because of untracked failed deployments
 	if contractID, ok := dl.NodeDeploymentID[dl.NodeID]; ok && contractID != 0 {
 		dl.ContractID = contractID
-		if !workloads.Contains(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID) {
-			d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID] = append(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID)
-		}
-		updateNetworkUsedIPs(&network, dl)
-		d.tfPluginClient.State.Networks[dl.NetworkName] = network
+		d.tfPluginClient.State.StoreContractIDs(dl.NodeID, []uint64{dl.ContractID})
 	}
 
 	return err
 }
 
 // BatchDeploy deploys multiple deployments using the deployer
 func (d *DeploymentDeployer) BatchDeploy(ctx context.Context, dls []*workloads.Deployment) error {
-	newDeployments := make(map[uint32][]gridtypes.Deployment)
 	newDeploymentsSolutionProvider := make(map[uint32][]*uint64)
-	networkUsedIPs := make(map[string]map[uint32][]byte)
 	var multiErr error
-	for _, dl := range dls {
-		if err := d.Validate(ctx, dl); err != nil {
-			multiErr = multierror.Append(multiErr, fmt.Errorf("invalid deployment %s: %w", dl.Name, err))
-			continue
-		}
-		if _, ok := networkUsedIPs[dl.NetworkName]; !ok {
-			networkUsedIPs[dl.NetworkName] = make(map[uint32][]byte)
-			network := d.tfPluginClient.State.Networks.GetNetwork(dl.NetworkName)
-			networkUsedIPs[dl.NetworkName][dl.NodeID] = network.GetUsedNetworkHostIDs(dl.NodeID)
-		}
-		generatedDls, usedHosts, err := d.GenerateVersionlessDeployments(ctx, dl, networkUsedIPs[dl.NetworkName][dl.NodeID])
-		if err != nil {
-			multiErr = multierror.Append(multiErr, fmt.Errorf("could not generate deployments data for deployment %s: %w", dl.Name, err))
-			continue
-		}
-		networkUsedIPs[dl.NetworkName][dl.NodeID] = usedHosts
 
-		for nodeID, generatedDl := range generatedDls {
-			if _, ok := newDeployments[nodeID]; !ok {
-				newDeploymentsSolutionProvider[nodeID] = []*uint64{dl.SolutionProvider}
-				newDeployments[nodeID] = []gridtypes.Deployment{generatedDl}
-				continue
-			}
-			newDeployments[nodeID] = append(newDeployments[nodeID], generatedDl)
-			newDeploymentsSolutionProvider[nodeID] = append(newDeploymentsSolutionProvider[nodeID], dl.SolutionProvider)
-		}
+	if err := d.Validate(ctx, dls); err != nil {
+		multiErr = multierror.Append(multiErr, fmt.Errorf("invalid deployments: %w", err))
+	}
+
+	newDeployments, err := d.GenerateVersionlessDeployments(ctx, dls)
+	if err != nil {
+		multiErr = multierror.Append(multiErr, fmt.Errorf("could not generate grid deployments: %w", err))
+	}
+
+	if len(newDeployments) == 0 {
+		return errors.Wrap(err, "failed to generate the grid deployments")
```

Comment:
why should I continue with no deployments?

Reply:
no no you shouldn't continue, I meant if err here is nil it will also return nil, and multiErr should have both errors from validation and from generating deployments so it would be better to use it.